### PR TITLE
basic write support for PXLC, PXSC, PXRS, and PXSC2

### PR DESCRIPTION
**Pull request by tmakatos**
Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

File: lib/pci_caps.c [hunk 1]
```diff
@@ -292,6 +292,44 @@ handle_px_pxdc_write(vfu_ctx_t *vfu_ctx, struct pxcap *px,
...
+static int
+handle_px_pxlc_write(vfu_ctx_t *vfu_ctx UNUSED, struct pxcap *px,
+                     const union pxlc *const p)
+{
+    assert(px != NULL);
+    assert(p != NULL);
+
+    /* FIXME RCB only writeable for endpoints and bridges */
+    /* FIXME must preserve reserved bits */
+    px->pxlc.raw = p->raw;
```
Comment (jlevon) [inline lib/pci_caps.c:304]: I think it's probably better to just ignore the write?
That way at least e.g. .rl behaviour is preserved.

File: lib/pci_caps.c [hunk 2]
```diff
@@ -292,6 +292,44 @@ handle_px_pxdc_write(vfu_ctx_t *vfu_ctx, struct pxcap *px,
...
+
+static int
+handle_px_pxsc_write(vfu_ctx_t *vfu_ctx UNUSED, struct pxcap *px,
+                     const struct pxsc *const p)
+{
+    assert(px != NULL);
+    assert(p != NULL);
+
+    /* FIXME */
+    px->pxsc.stuff = p->stuff;
```
Comment (jlevon) [inline lib/pci_caps.c:316]: ditto?

File: lib/pci_caps.c [hunk 3]
```diff
@@ -292,6 +292,44 @@ handle_px_pxdc_write(vfu_ctx_t *vfu_ctx, struct pxcap *px,
...
+}
+
+static int
+handle_px_pxrc_write(vfu_ctx_t *vfu_ctx UNUSED, struct pxcap *px,
+                     const struct pxrc *const p)
+{
+    assert(px != NULL);
+    assert(p != NULL);
+
+    /* FIXME  */
```
Comment (jlevon) [inline lib/pci_caps.c:327]: skip this line?

File: lib/pci_caps.c [hunk 4]
```diff
@@ -292,6 +292,44 @@ handle_px_pxdc_write(vfu_ctx_t *vfu_ctx, struct pxcap *px,
...
+    /* FIXME must preserve reserved bits */
+    px->pxrc.stuff = p->stuff;
```
Comment (jlevon) [inline lib/pci_caps.c:329]: this matters a bit less I suppose, as it's root port specific.

**Review by jlevon — CHANGES_REQUESTED** [review]

File: lib/pci_caps.c (hunk 1 above)
Comment (tmakatos) [reply lib/pci_caps.c:304]: sure

File: lib/pci_caps.c (hunk 2 above)
Comment (tmakatos) [reply lib/pci_caps.c:316]: sure

File: lib/pci_caps.c (hunk 3 above)
Comment (tmakatos) [reply lib/pci_caps.c:327]: ok

File: lib/pci_caps.c (hunk 4 above)
Comment (tmakatos) [reply lib/pci_caps.c:329]: ack

**Review by jlevon — APPROVED** [review]
thanks

**Pull request merged**